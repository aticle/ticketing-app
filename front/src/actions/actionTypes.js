// @flow
export default {
    CREATE_NEW_TICKET: 'CREATE_NEW_TICKET',
    GET_ALL_TICKETS: 'GET_ALL_TICKETS',
    DELETE_TICKET: 'DELETE_TICKET',
    EDIT_TICKET: 'EDIT_TICKET'
}