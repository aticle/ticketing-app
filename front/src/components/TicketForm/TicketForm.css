.create-ticket-form {
    padding: 20px;
    display: flex;
    align-items: center;
    flex-direction: column;
}

.create-ticket-form>div {
    margin: 5px 0;
    display: flex;
    width: 300px;
    max-width: 100%;
    text-align: left;
}

.create-ticket-form button[type="submit"] {
    margin-top: 65px;
}