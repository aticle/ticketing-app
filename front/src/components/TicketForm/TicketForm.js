// @flow
import React, { Component } from 'react';
import { connect, type Dispatch } from 'react-redux';
import { TextField, Button } from '@material-ui/core';

import { type TicketState, Status } from '../Ticket/Ticket';
import { createTicket } from '../../actions/ticketAction';
import type { State } from '../../reducers';

import './TicketForm.css';

type Props = {
    createTicket: Dispatch,
    tickets: Array<TicketState>,
    ticket: TicketState
}

class TicketForm extends Component<Props, TicketState> {
    state = {
        status: Status.OPEN,
        title: '',
        description: ''
    };

    handleChange = (name: string) => (e: SyntheticEvent<HTMLInputElement>) => {
        this.setState({
            [name]: e.currentTarget.value
        });
    }

    handleSubmit = (e: Event) => {
        e.preventDefault();

        const ticket: TicketState = Object.assign({}, this.state);
        this.props.createTicket(ticket);

        this.setState({
            title: '',
            description: ''
        });
    }

    render() {
        const { title, description } = this.state;

        return (
            <form
                className="create-ticket-form"
                onSubmit={this.handleSubmit}
            >
                <h2>Create Ticket</h2>
                <TextField
                    label="Title"
                    id="title"
                    value={title}
                    required
                    onChange={this.handleChange('title')}
                ></TextField>
                <TextField
                    label="Description"
                    id="description"
                    value={description}
                    required
                    onChange={this.handleChange('description')}
                ></TextField>
                <Button
                    id="submit"
                    variant='contained'
                    type="submit"
                >SUBMIT</Button>
            </form>
        );
    };
}
export const mapStateToProps = (state: State) => {
    return {
        tickets: state.tickets
    }
};
export const mapDispatchToProps = (dispatch: Dispatch) => {
    return {
        createTicket: (ticket: TicketState) => dispatch(createTicket(ticket))
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(TicketForm);
