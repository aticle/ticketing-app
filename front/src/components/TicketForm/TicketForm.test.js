import React from 'react';
import { shallow } from 'enzyme';
import actionTypes from '../../actions/actionTypes'
import TicketForm, { mapStateToProps, mapDispatchToProps } from './TicketForm';

describe('<TicketForm />', () => {
    it('renders correctly', () => {
        const createTicket = shallow(<TicketForm.WrappedComponent />);
        expect(createTicket).toMatchSnapshot();
    });

    it('handles correctly onChange', () => {
        const createTicket = shallow(<TicketForm.WrappedComponent />);

        // ID
        const idValue = 5;
        const idEvent = {
            currentTarget: {
                value: idValue.toString(),
                valueAsNumber: idValue,
                type: "number"
            }
        };
        createTicket.find("#id").simulate('change', idEvent);
        expect(createTicket.state().id).toEqual(idValue);

        // title
        const titleValue = "Title text";
        const titleEvent = {
            currentTarget: {
                value: titleValue,
                valueAsNumber: Number(titleValue),
                type: "text"
            }
        };
        createTicket.find("#title").simulate('change', titleEvent);
        expect(createTicket.state().title).toEqual(titleValue);

        // description
        const descriptionValue = "Description text";
        const descriptionEvent = {
            currentTarget: {
                value: descriptionValue,
                valueAsNumber: Number(descriptionValue),
                type: "text"
            }
        };
        createTicket.find("#description").simulate('change', descriptionEvent);
        expect(createTicket.state().description).toEqual(descriptionValue);
    });

    it('correctly mapDispatchToProps', () => {
        const ticket = {
            id: 6,
            title: 'bla',
            description: 'bla bla bla',
            status: 'OPEN'
        };
        const dispatch = jest.fn();

        mapDispatchToProps(dispatch).createTicket(ticket);
        expect(dispatch.mock.calls[0][0]).toEqual({
            type: actionTypes.CREATE_NEW_TICKET,
            payload: {
                request: {
                    data: ticket,
                    method: "post",
                    url: "/tickets/create"
                }
            }
        });
    });


    it('correctly mapStateToProps', () => {
        const ticket = {
            id: 6,
            title: 'bla',
            description: 'bla bla bla',
            status: 'OPEN'
        };

        expect(mapStateToProps({ tickets: [ticket] })).toHaveProperty('tickets', [ticket]);
    });

    it('handles correctly onSubmit', () => {
        const event = { preventDefault: () => { } };
        const dispatch = jest.fn();
        const props = { ...mapDispatchToProps(dispatch), ...mapStateToProps({ tickets: [] }) };
        const createTicketWrap = shallow(<TicketForm.WrappedComponent {...props} />);
        const ticket = {
            id: 6,
            title: 'bla',
            description: 'bla bla bla',
            status: 'OPEN'
        };
        expect(createTicketWrap.find('form')).toHaveLength(1);
        createTicketWrap.setState({
            ...createTicketWrap.state,
            id: 6,
            title: 'bla',
            description: 'bla bla bla'
        });
        createTicketWrap.find('form').simulate('submit', event);
        expect(createTicketWrap.state()).toEqual({
            id: ticket.id + 1,
            title: '',
            description: '',
            status: 'OPEN'
        });
    });
}); 