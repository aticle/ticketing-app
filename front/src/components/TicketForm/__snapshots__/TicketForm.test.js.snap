// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketForm /> renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TicketForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Create Ticket
        </h2>,
        <TextField
          id="title"
          label="Title"
          onChange={[Function]}
          required={true}
          select={false}
          value=""
          variant="standard"
        />,
        <TextField
          id="description"
          label="Description"
          onChange={[Function]}
          required={true}
          select={false}
          value=""
          variant="standard"
        />,
        <WithStyles(Button)
          id="submit"
          type="submit"
          variant="contained"
        >
          SUBMIT
        </WithStyles(Button)>,
      ],
      "className": "create-ticket-form",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Create Ticket",
        },
        "ref": null,
        "rendered": "Create Ticket",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "id": "title",
          "label": "Title",
          "onChange": [Function],
          "required": true,
          "select": false,
          "value": "",
          "variant": "standard",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "id": "description",
          "label": "Description",
          "onChange": [Function],
          "required": true,
          "select": false,
          "value": "",
          "variant": "standard",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "SUBMIT",
          "id": "submit",
          "type": "submit",
          "variant": "contained",
        },
        "ref": null,
        "rendered": "SUBMIT",
        "type": [Function],
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Create Ticket
          </h2>,
          <TextField
            id="title"
            label="Title"
            onChange={[Function]}
            required={true}
            select={false}
            value=""
            variant="standard"
          />,
          <TextField
            id="description"
            label="Description"
            onChange={[Function]}
            required={true}
            select={false}
            value=""
            variant="standard"
          />,
          <WithStyles(Button)
            id="submit"
            type="submit"
            variant="contained"
          >
            SUBMIT
          </WithStyles(Button)>,
        ],
        "className": "create-ticket-form",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Create Ticket",
          },
          "ref": null,
          "rendered": "Create Ticket",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "id": "title",
            "label": "Title",
            "onChange": [Function],
            "required": true,
            "select": false,
            "value": "",
            "variant": "standard",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "id": "description",
            "label": "Description",
            "onChange": [Function],
            "required": true,
            "select": false,
            "value": "",
            "variant": "standard",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "SUBMIT",
            "id": "submit",
            "type": "submit",
            "variant": "contained",
          },
          "ref": null,
          "rendered": "SUBMIT",
          "type": [Function],
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenThreeAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
